### ISSUE TITLE
Docker?

### ISSUE THREAD

**Issue by aldycool** [Grokzen/redis-py-cluster#179]
Hi @grokzen, I'm not sure if this is off-topic in regards to the redis-cluster-py, but this has been nagging me for a week now. In the Redis Cluster Specification, it said that Docker only supported if using host network only, means that Redis Cluster cannot be deployed using Docker in multiple servers. Is this true? Or do you know a way to solve this? I already read the redis github issue about this matter, looks like they want to support announce-ip (so they can use the internal Docker Container IP) but this have not been implemented yet. I'm very sorry if this is out of place, but I'm really curious whether you have valuable thoughts about this. Thanks!

**Comment by Grokzen** [comment]
@aldycool Hi. Here is my 2 cents on the docker problem.

The TLDR version is the following few points
 - Yes docker is supported
 - This problem is something that redis & docker them self has to solve (This is my personal opinion)

The long answer.

Okey so lets start with some internals on how the client works internally and how cluster discovery works.

When you create a client instance it will issue a command to one of the nodes that you specify in the list of  `startup_nodes` that you send into the client class (See nodemanager.py#L178). That command will return a data structure that describes the structure of the cluster. It defines what nodes that exists, master/slave relations, how to reach all them and the slots layout. This data is then used to populate the internal data structures used by the client so that when you issue a command, it knows where to try to send each command into the cluster.

So lets talk about how the cluster is setup and why this is what is "breaking" the docker support. When you are creating a cluster and you want all the nodes to meet and start to build the cluster you probably execute this command `ruby /redis/src/redis-trib.rb create --replicas 1 127.0.0.1:7000 127.0.0.1:7001 ...` for all nodes so that they will join up in a working cluster. The problem with this approach is that redis is kinda stupid/simple where it will blindly look at the ip addresses that you specefied as arguments to that command and use them as the locations where all the other cluster nodes is. It do not try to do some lookup of your public ip or anything in that sense becuase in one way it is not up to redis to understand or try to figure out what exact scenario you want. Becuase there is 1000+ ways of setting up a cluster, redis should know about all cases and behave exactly like you want out of the box. This is not something that you can expect redis to do becuase it is not reasonable to put that kind of logic into any system. So, the solution has to be done on your side and exactly how you setup your cluster.

So in order to fix the problem with docker, you have to fix the parameters you send into `redis-trib.rb` becuase that is the only way to solve the problem for real in all cases. So there is a few different kinds of solutions to this problem that you can use to solve the IP problem, but it all depends on your specific setup and how you operate your cluster.

If you look for example here https://github.com/Grokzen/docker-redis-cluster/blob/master/docker-data/start.sh#L4 and how i do it in my docker image, i use a small `ifconfig | grep ...`command to determine the public IP address that docker uses on my own machine in order to determine the IP address that i should send into `redis-trib.rb`. What this does is that when you later in your client sends the cluster `CLUSTER SLOTS` command, it will return the IP addresses of all nodes no matter in what container i have them inside and my client that i run outside docker (on my host machine) will be able to connect to all nodes in the cluster. The downside of this approach is that it do not work if you want to run the client instance on another machine over the network. The reason for this is that if your cluster tracks the IP address `172.20.0.2` and the machine you host it on has `192.168.0.123` and the machine that runs the client has ip `192.168.0.23` then the .23 machine will never understand what/where/how to reach the address `172.20.0.2` because that is running locally on the .123 machine.

The only known way to solve the case above is to somehow in your setup find the IP address that you use on your host and send that into `redis-trib.py` so that it can discover all the nodes on all other machines you run. This problem however is not something that is easily solved in a general way becuase you have to adapt to your own infrastructure and needs.

What you can do is to use DNS records to help track the IP addresses of machines and you send that into a shell script that does a lookup of the IP and that IP can be the public address of your host machine so that even from inside your docker container you can reach to another docker container on another host just based on public ip of the other host machine.

One thing you can't do right now (unless it was added recently) is to send DNS names into redis-trib.rb instead of having to send in IP addresses and then you only need to have the DNS working proerly and redis would handle looking it up and trying to connect to each node and then the problem would go away rather easily (i think atleast O.o)

**Comment by Grokzen** [comment]
Closing this one because it is not a real issue per say. We can still continue talk about it in this issue or you can ping me over at gitter.im (free chat, see link in readme)

**Comment by ashhar-hasan** [comment]
It looks like we are running into this issue. How can we be sure that this is the issue and not something else? The script seems to hang at rc.Ping() and sometimes rc.StrictRedisCluster().

Here's a stack trace:

```log
File "import_size_fit_accuracy_data_into_redis.py", line 86, in <module>
    start()
  File "import_size_fit_accuracy_data_into_redis.py", line 83, in start
    init()
  File "import_size_fit_accuracy_data_into_redis.py", line 11, in init
    initRedisCon()
  File "import_size_fit_accuracy_data_into_redis.py", line 42, in initRedisCon
    rc = rediscluster.StrictRedisCluster(startup_nodes=startup_nodes,decode_responses=True)
  File "/usr/local/lib/python2.7/dist-packages/rediscluster/client.py", line 181, in __init__
    **kwargs
  File "/usr/local/lib/python2.7/dist-packages/rediscluster/connection.py", line 141, in __init__
    self.nodes.initialize()
  File "/usr/local/lib/python2.7/dist-packages/rediscluster/nodemanager.py", line 228, in initialize
    need_full_slots_coverage = self.cluster_require_full_coverage(nodes_cache)
  File "/usr/local/lib/python2.7/dist-packages/rediscluster/nodemanager.py", line 270, in cluster_require_full_coverage
    return any(node_require_full_coverage(node) for node in nodes.values())
  File "/usr/local/lib/python2.7/dist-packages/rediscluster/nodemanager.py", line 270, in <genexpr>
    return any(node_require_full_coverage(node) for node in nodes.values())
  File "/usr/local/lib/python2.7/dist-packages/rediscluster/nodemanager.py", line 267, in node_require_full_coverage
    return "yes" in r_node.config_get("cluster-require-full-coverage").values()
  File "/usr/local/lib/python2.7/dist-packages/redis/client.py", line 715, in config_get
    return self.execute_command('CONFIG GET', pattern)
  File "/usr/local/lib/python2.7/dist-packages/redis/client.py", line 667, in execute_command
    connection.send_command(*args)
  File "/usr/local/lib/python2.7/dist-packages/redis/connection.py", line 610, in send_command
    self.send_packed_command(self.pack_command(*args))
  File "/usr/local/lib/python2.7/dist-packages/redis/connection.py", line 585, in send_packed_command
    self.connect()
  File "/usr/local/lib/python2.7/dist-packages/redis/connection.py", line 484, in connect
    sock = self._connect()
  File "/usr/local/lib/python2.7/dist-packages/redis/connection.py", line 529, in _connect
    sock.connect(socket_address)
  File "/usr/lib/python2.7/socket.py", line 224, in meth
    return getattr(self._sock,name)(*args)
KeyboardInterrupt
```

**Comment by Grokzen** [comment]
@ashhar-hasan do that happen every time or just some of the time?

**Comment by ashhar-hasan** [comment]
Everytime, the only difference is in whether is stops at rc.StrictRedisCluster() or at rc.Ping().

**Comment by ashhar-hasan** [comment]
Running this locally on a redis cluster poses no issues, only when docker is involved this happens. I will test it another environment and see I can reproduce there.

The important part is that the script was working fine until some time ago and that's what makes me think this is an issue caused by a config change in the redis cluster.

How can I know what IP is redis using to address nodes? Using CLUSTER NODES, CLUSTER INFO?

**Comment by Grokzen** [comment]
Cluster discovery is done here with `CLUSTER SLOTS` as used command https://github.com/Grokzen/redis-py-cluster/blob/unstable/rediscluster/nodemanager.py#L178

If you issue that command to your redis cluster inside docker, you will get the IP addresses that this client uses to connect to the nodes. If your client can't reach the IP addresses then it will fail with similar exceptions you have above. This is a classic problem with docker installations that either your network is bad, or you cluster on the inside is bad.
